gdpr_assist/models.py

```diff
@@ -246,33 +278,278 @@ def _cast_class(cls, model, privacy_meta):
+                actual_anon_log_line_start['model_name'],
+                actual_anon_log_line_start['target_pk'],
+                user,
+                actual_anon_log_line_start["log_time"].strftime("%Y-%m-%d %H:%M:%S")
+            )
+            indent_level = 0
+
+            def tabify(x):
+                return "\t" * x if x > 0 else ""
+
+            for l in lines:
+                if l["event"] == EventLog.EVENT_RECURSIVE_START:
+                    res += "%sStarting recursive for %s #%s.\n" % (
+                    tabify(indent_level), l['model_name'], l['target_pk'])
+                    indent_level += 1
+
+                elif l["event"] == EventLog.EVENT_RECURSIVE_END:
+                    indent_level -= 1
+                    res += "%sEnding recursive for %s #%s.\n" % (tabify(indent_level), l['model_name'], l['target_pk'])
+
+                elif l["event"] == EventLog.EVENT_ALREADY_ANONYMISED:
+                    res += "%s%s #%s already anonymised.\n" % (tabify(indent_level), l['model_name'], l['target_pk'])
+
+                else:
+                    res += "%s%s #%s flat fields anonymised.\n" % (
+                    tabify(indent_level), l['model_name'], l['target_pk'])
+
+            return res
+        else:
+            return "%s is anonymised, but does not have matching logs." % self
+
+    @classmethod
+    def get_anonymization_tree(cls, prefix="", doprint=False, objs=[]):
+        """
+        Print the result of the nesting defined above for a given model.
+        Useful for sanity and loop checking. Example output:
+
+        Class:
+        |-> outcome_assessments = (OutcomeAssessment [set_field]):
+            |-> score
+            |-> comment
+        |-> class_polls = (ClassPoll [set_field]):
+            |-> poll = (Poll [fk]):
+                |-> title
+                |-> question
+                |-> pollchoice_set = (PollChoice [set_field]):
+                    |-> choice
+                |-> pollsession_set = (PollSession [set_field]):
+                    |-> outcome_assessments = (OutcomeAssessment [set_field]):
+                        |-> score
+                        |-> comment
+                    |-> pollresponse_set = (PollResponse [set_field]):
+                        |-> response
+
+        """
+        ABANDON_AFTER_N_LEVELS = 10
+        BASE_PREFIX = "    "
+        if len(prefix) > len(BASE_PREFIX) * ABANDON_AFTER_N_LEVELS:
+            return "ERROR: shouldn't go %s levels deep, check for loops." % ABANDON_AFTER_N_LEVELS
+        res = ''
+        if doprint:
+            res += cls.__name__ + ":\n"
+
+        privacy_meta_model = cls._privacy_meta
+
+        flat_fields = privacy_meta_model.fields
+        if flat_fields:
+            for f in flat_fields:
+                res += "%s|-> %s\n" % (prefix, f)
+
+        for fk in privacy_meta_model.fk_fields:
+
+            try:
+                f = getattr(cls, fk)
```

Comment:
What's `f`? A model field? I know it's a draft but it would really help readability to use meaningful variable names instead of one to three letters names.

Reply:
sortof. see the next few lines.

Either it's something we can do `f.field` or `f.rel` on. There a bunch of types that can have those, and I personally do not want to manually check for all of them. This seemed more general.

Basically, you just specify the field, and it doesnt matter if it is a regular FK, a reverse FK, a Relationship of some kind.

So, `f` is the regular of doing `cls.{fk}`. Do you think `field` gives you more info? I don't think much more.